File: libs/server/persistent/src/persistent/filesystem/workspace-logical-entities.ts
import { randomUUID } from 'node:crypto';
import { mkdir, rm, writeFile } from 'node:fs/promises';
import { join } from 'node:path';
import {
  DomainError,
  LogicalEntity,
  type LogicalEntityDescription,
  normalizeSubType,
  parseLogicalEntityType,
  Ref,
  type WorkspaceLogicalEntities,
} from '@evidence/server-domain';
import { stringify } from 'yaml';
import { EntityList } from '../database';
import {
  fileTimestamp,
  listYamlFiles,
  optionalString,
  readYamlRecord,
  requiredString,
  type YamlRecord,
} from './model-files';

interface EntityRecord {
  entity: LogicalEntity;
  parent: string | null;
  path: string;
}

export class FileWorkspaceLogicalEntities
  extends EntityList<LogicalEntity>
  implements WorkspaceLogicalEntities
{
  private readonly entitiesDirectory: string;

  constructor(
    private readonly workspaceId: string,
    evidenceRoot: string,
  ) {
    super();
    this.entitiesDirectory = join(evidenceRoot, 'entities');
  }

  override async size(): Promise<number> {
    return (await this.load()).length;
  }

  protected override async findEntities(
    from: number,
    to: number,
  ): Promise<LogicalEntity[]> {
    return (await this.load()).slice(from, to).map((record) => record.entity);
  }

  protected override async findEntity(
    id: string,
  ): Promise<LogicalEntity | null> {
    return (await this.findRecord(id))?.entity ?? null;
  }

  async add(desc: LogicalEntityDescription): Promise<LogicalEntity> {
    const name = normalizeName(desc.name);
    const id = await this.availableId(name);
    const path = join(this.entitiesDirectory, `${id}.yaml`);
    return this.writeEntity(path, id, { ...desc, name }, null);
  }

  async update(
    entityId: string,
    desc: LogicalEntityDescription,
  ): Promise<LogicalEntity> {
    const record = await this.findRecord(entityId);
    if (!record) {
      throw DomainError.notFound(`logical entity ${entityId} not found`);
    }
    return this.writeEntity(
      record.path,
      entityId,
      { ...desc, name: normalizeName(desc.name) },
      record.parent,
    );
  }

  async delete(entityId: string): Promise<void> {
    const record = await this.findRecord(entityId);
    if (!record) {
      throw DomainError.notFound(`logical entity ${entityId} not found`);
    }
    try {
      await rm(record.path);
    } catch (error) {
      throw DomainError.internal(
        `delete logical entity file ${record.path}: ${errorMessage(error)}`,
      );
    }
  }

  async list(
    page: number,
    pageSize: number,
  ): Promise<[LogicalEntity[], number]> {
    rejectInvalidPage(page, pageSize);
    const records = await this.load();
    const offset = (page - 1) * pageSize;
    return [
      records.slice(offset, offset + pageSize).map((record) => record.entity),
      records.length,
    ];
  }

  private async load(): Promise<EntityRecord[]> {
    const paths = await listYamlFiles(this.entitiesDirectory);
    const records = await Promise.all(
      paths.map((path) => this.readEntity(path)),
    );
    records.sort((left, right) => {
      const leftDescription = left.entity.description();
      const rightDescription = right.entity.description();
      return (
        leftDescription.name.localeCompare(rightDescription.name) ||
        left.entity.identity().localeCompare(right.entity.identity())
      );
    });
    return records;
  }

  private async findRecord(id: string): Promise<EntityRecord | null> {
    return (
      (await this.load()).find((record) => record.entity.identity() === id) ??
      null
    );
  }

  private async readEntity(path: string): Promise<EntityRecord> {
    const document = await readYamlRecord(path, 'logical entity');
    const id = requiredString(document, 'id', path, 'logical entity');
    const name = requiredString(document, 'name', path, 'logical entity');
    const type = parseLogicalEntityType(
      requiredString(document, 'type', path, 'logical entity'),
    );
    const subType = normalizeSubType(
      type,
      optionalString(document.subType ?? document.sub_type),
    );
    const parent = optionalString(document.parent);
    const timestamp = await fileTimestamp(path);

    return {
      path,
      parent,
      entity: new LogicalEntity(id, {
        workspace: new Ref(this.workspaceId),
        type,
        subType,
        name,
        label: optionalString(document.label),
        description: optionalContent(document.content ?? document.description),
        attributes: entityAttributes(document),
        createdAt: timestamp,
        updatedAt: timestamp,
      }),
    };
  }

  private async writeEntity(
    path: string,
    id: string,
    desc: LogicalEntityDescription,
    parent: string | null,
  ): Promise<LogicalEntity> {
    const subType = normalizeSubType(desc.type, desc.subType);
    const document: YamlRecord = {
      id,
      name: desc.name,
      ...(desc.label ? { label: desc.label } : {}),
      type: desc.type,
      ...(subType ? { subType } : {}),
      ...(parent ? { parent } : {}),
      ...(desc.description ? { description: desc.description } : {}),
      ...(desc.attributes.length > 0 ? { attributes: desc.attributes } : {}),
    };
    try {
      await mkdir(this.entitiesDirectory, { recursive: true });
      await writeFile(path, stringify(document, { lineWidth: 0 }), 'utf8');
    } catch (error) {
      throw DomainError.internal(
        `write logical entity file ${path}: ${errorMessage(error)}`,
      );
    }
    return (await this.readEntity(path)).entity;
  }

  private async availableId(name: string): Promise<string> {
    const base = normalizeIdentifier(name) ?? randomUUID();
    if (!(await this.findRecord(base))) {
      return base;
    }

    for (;;) {
      const candidate = `${base}_${randomUUID().split('-')[0]}`;
      if (!(await this.findRecord(candidate))) {
        return candidate;
      }
    }
  }
}

function normalizeName(name: string): string {
  const normalized = name.trim();
  if (!normalized) {
    throw DomainError.validation('logical entity name must not be empty');
  }
  return normalized;
}

function normalizeIdentifier(value: string): string | null {
  const normalized = value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '');
  return normalized || null;
}

function optionalContent(value: unknown): string | null {
  return typeof value === 'string' && value.trim().length > 0 ? value : null;
}

function entityAttributes(
  document: YamlRecord,
): LogicalEntityDescription['attributes'] {
  if (!Array.isArray(document.attributes)) {
    return [];
  }
  return document.attributes.filter(isEntityAttribute);
}

function isEntityAttribute(
  value: unknown,
): value is LogicalEntityDescription['attributes'][number] {
  return Boolean(
    value &&
      typeof value === 'object' &&
      'id' in value &&
      typeof value.id === 'string' &&
      'name' in value &&
      typeof value.name === 'string',
  );
}

function rejectInvalidPage(page: number, pageSize: number): void {
  if (page <= 0 || pageSize <= 0) {
    throw DomainError.validation('page and pageSize must be greater than 0');
  }
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: libs/server-java/persistent/src/main/java/reengineering/ddd/evidence/persistent/filesystem/WorkspaceLogicalEntities.java
package reengineering.ddd.evidence.persistent.filesystem;

import io.github.jayclock.smartdomain.core.Ref;
import io.github.jayclock.smartdomain.mybatis.AssociationMapping;
import io.github.jayclock.smartdomain.mybatis.database.EntityList;
import jakarta.inject.Inject;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import reengineering.ddd.evidence.domain.DomainException;
import reengineering.ddd.evidence.domain.model.LogicalEntity;
import reengineering.ddd.evidence.domain.model.Workspace;
import reengineering.ddd.evidence.persistent.mappers.WorkspacesMapper;

@AssociationMapping(
    entity = Workspace.class,
    field = "logicalEntities",
    parentIdField = "workspaceId")
public final class WorkspaceLogicalEntities extends EntityList<String, LogicalEntity>
    implements Workspace.LogicalEntities {
  private String workspaceId;
  @Inject private WorkspacesMapper workspaces;

  public WorkspaceLogicalEntities() {}

  WorkspaceLogicalEntities(String workspaceId, WorkspacesMapper workspaces) {
    this.workspaceId = workspaceId;
    this.workspaces = workspaces;
  }

  @Override
  protected List<LogicalEntity> findEntities(int from, int to) {
    List<EntityRecord> records = load();
    return records.subList(Math.min(from, records.size()), Math.min(to, records.size())).stream()
        .map(EntityRecord::entity)
        .toList();
  }

  @Override
  protected LogicalEntity findEntity(String id) {
    EntityRecord record = findRecord(id);
    return record == null ? null : record.entity();
  }

  @Override
  public int size() {
    return load().size();
  }

  @Override
  public LogicalEntity add(LogicalEntity.Description description) {
    validateWorkspace(description.workspace().id());
    String name = normalizeName(description.name());
    String id = availableId(name);
    Path path = entitiesDirectory().resolve(id + ".yaml");
    return write(path, id, copyWithName(description, name), null, true);
  }

  @Override
  public LogicalEntity update(String entityId, LogicalEntity.Description description) {
    validateWorkspace(description.workspace().id());
    EntityRecord current = findRecord(entityId);
    if (current == null) {
      throw DomainException.notFound("logical entity " + entityId + " not found");
    }
    return write(
        current.path(),
        entityId,
        copyWithName(description, normalizeName(description.name())),
        current.parent(),
        false);
  }

  @Override
  public void remove(String entityId) {
    EntityRecord current = findRecord(entityId);
    if (current == null) {
      throw DomainException.notFound("logical entity " + entityId + " not found");
    }
    ModelFiles.delete(current.path(), "logical entity");
  }

  @Override
  public Workspace.Page<LogicalEntity> list(int page, int pageSize) {
    validatePage(page, pageSize);
    List<EntityRecord> records = load();
    int from = Math.min((page - 1) * pageSize, records.size());
    int to = Math.min(from + pageSize, records.size());
    return new Workspace.Page<>(
        records.subList(from, to).stream().map(EntityRecord::entity).toList(), records.size());
  }

  private List<EntityRecord> load() {
    return ModelFiles.listYamlFiles(entitiesDirectory()).stream()
        .map(this::read)
        .sorted(
            Comparator.comparing((EntityRecord record) -> record.entity().getDescription().name())
                .thenComparing(record -> record.entity().getIdentity()))
        .toList();
  }

  private EntityRecord findRecord(String id) {
    return load().stream()
        .filter(record -> record.entity().getIdentity().equals(id))
        .findFirst()
        .orElse(null);
  }

  private EntityRecord read(Path path) {
    Map<String, Object> document = ModelFiles.readYaml(path, "logical entity");
    String id = ModelFiles.requiredString(document, "id", path, "logical entity");
    String name = ModelFiles.requiredString(document, "name", path, "logical entity");
    LogicalEntity.Type type =
        LogicalEntity.parseType(
            ModelFiles.requiredString(document, "type", path, "logical entity"));
    String subType =
        LogicalEntity.normalizeSubType(
            type,
            ModelFiles.optionalString(
                document.get("subType") != null
                    ? document.get("subType")
                    : document.get("sub_type")));
    String parent = ModelFiles.optionalString(document.get("parent"));
    Object descriptionValue =
        document.get("content") != null ? document.get("content") : document.get("description");
    String description =
        descriptionValue instanceof String text && !text.trim().isEmpty() ? text : null;
    var timestamp = ModelFiles.timestamp(path);
    return new EntityRecord(
        new LogicalEntity(
            id,
            new LogicalEntity.Description(
                new Ref<>(workspaceId),
                type,
                subType,
                name,
                ModelFiles.optionalString(document.get("label")),
                description,
                attributes(document.get("attributes")),
                timestamp,
                timestamp)),
        parent,
        path);
  }

  private LogicalEntity write(
      Path path, String id, LogicalEntity.Description description, String parent, boolean create) {
    String subType = LogicalEntity.normalizeSubType(description.type(), description.subType());
    Map<String, Object> document = new LinkedHashMap<>();
    document.put("id", id);
    document.put("name", description.name());
    if (description.label() != null) document.put("label", description.label());
    document.put("type", description.type().name());
    if (subType != null) document.put("subType", subType);
    if (parent != null) document.put("parent", parent);
    if (description.description() != null) {
      document.put("description", description.description());
    }
    if (!description.attributes().isEmpty()) {
      document.put("attributes", description.attributes());
    }
    if (create) {
      ModelFiles.writeNewYaml(path, document, "logical entity");
    } else {
      ModelFiles.replaceYaml(path, document, "logical entity");
    }
    return read(path).entity();
  }

  private String availableId(String name) {
    String normalized = normalizeIdentifier(name);
    String base = normalized == null ? UUID.randomUUID().toString() : normalized;
    String candidate = base;
    while (findRecord(candidate) != null
        || Files.exists(entitiesDirectory().resolve(candidate + ".yaml"))) {
      candidate = base + "_" + UUID.randomUUID().toString().substring(0, 8);
    }
    return candidate;
  }

  private Path entitiesDirectory() {
    String root = workspaces.findModelRoot(workspaceId);
    if (root == null) {
      throw DomainException.notFound("workspace " + workspaceId + " not found");
    }
    return Path.of(root).toAbsolutePath().normalize().resolve("entities");
  }

  private void validateWorkspace(String requestedWorkspaceId) {
    if (!workspaceId.equals(requestedWorkspaceId)) {
      throw DomainException.validation(
          "logical entity workspace "
              + requestedWorkspaceId
              + " does not match scoped workspace "
              + workspaceId);
    }
  }

  private static LogicalEntity.Description copyWithName(
      LogicalEntity.Description description, String name) {
    return new LogicalEntity.Description(
        description.workspace(),
        description.type(),
        description.subType(),
        name,
        description.label(),
        description.description(),
        description.attributes(),
        description.createdAt(),
        description.updatedAt());
  }

  private static String normalizeName(String name) {
    String normalized = name == null ? "" : name.trim();
    if (normalized.isEmpty()) {
      throw DomainException.validation("logical entity name must not be empty");
    }
    return normalized;
  }

  private static String normalizeIdentifier(String value) {
    String normalized =
        value
            .toLowerCase(java.util.Locale.ROOT)
            .replaceAll("[^a-z0-9]+", "_")
            .replaceAll("^_+|_+$", "");
    return normalized.isEmpty() ? null : normalized;
  }

  private static List<LogicalEntity.Attribute> attributes(Object value) {
    if (!(value instanceof List<?> input)) return List.of();
    List<LogicalEntity.Attribute> attributes = new ArrayList<>();
    for (Object item : input) {
      if (!(item instanceof Map<?, ?> attribute)) continue;
      String id = ModelFiles.optionalString(attribute.get("id"));
      String name = ModelFiles.optionalString(attribute.get("name"));
      if (id == null || name == null) continue;
      attributes.add(
          new LogicalEntity.Attribute(
              id,
              name,
              ModelFiles.optionalString(attribute.get("label")),
              ModelFiles.optionalString(attribute.get("type")),
              ModelFiles.optionalString(attribute.get("description"))));
    }
    return List.copyOf(attributes);
  }

  private static void validatePage(int page, int pageSize) {
    if (page <= 0 || pageSize <= 0) {
      throw DomainException.validation("page and pageSize must be greater than 0");
    }
  }

  private record EntityRecord(LogicalEntity entity, String parent, Path path) {}
}
